### PR TITLE
Extend Builder to Support Validation of Required Attributes 

### PR DESCRIPTION
**Pull request by trentjeff**
* Adds builder attribute visitor.
* Uses Visitor for attribute validation of 'required' attributes (from ConfiguredOptions).
* For Singular, fixes the plurality of the verbs (addPickle instead of addPickles).

### CODE REVIEW

File: pico/builder/tests/builder/src/test/java/io/helidon/pico/builder/api/test/MyConfigBeanTest.java [hunk 1]
```diff
@@ -35,47 +35,47 @@ void manual() {
                    equalTo("MyConfigBeanManualImpl(name=null, enabled=false, port=0)"));
 
         val = MyConfigBeanManualImpl.toBuilder(val)
-                .name("jeff")
+                .name("test")
                 .enabled(true)
                 .port(80)
                 .build();
-        assertThat(val.toString(), equalTo("MyConfigBeanManualImpl(name=jeff, enabled=true, port=80)"));
+        assertThat(val.toString(), equalTo("MyConfigBeanManualImpl(name=test, enabled=true, port=80)"));
```
Comment (tomas-langer) [inline pico/builder/tests/builder/src/test/java/io/helidon/pico/builder/api/test/MyConfigBeanTest.java:42]: Doing assertions on `toString()` is not a good practice, as you make the `toString()` result a specification, where it should be a human readable string (but not specified or something you should ever depend on).
If you want to test the builder contains the correct values, use getters.
If you want to test the `toString()` contains all values, use multiple assertions with contains string (e.g. look just for the value, not for the whole format of toString())

TL;DR: do not test toString(), we may want to change it and it should not result in a failed test

Comment (tomas-langer) [reply pico/builder/tests/builder/src/test/java/io/helidon/pico/builder/api/test/MyConfigBeanTest.java:42]: The comment above (assertion on `toString()` is applicable to multiple tests)

File: pico/builder/tests/builder/src/test/java/io/helidon/pico/builder/api/test/UsageTest.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+
+class UsageTest {
+
+    @Test
+    void testIt() {
+        A a = DefaultA.builder()
+                .a("hello")
+                .build();
+        assertThat(a, notNullValue());
```
Comment (tomas-langer) [inline pico/builder/tests/builder/src/test/java/io/helidon/pico/builder/api/test/UsageTest.java:40]: I would expect assertions on getters here, to make sure they return the configured/default values

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: pico/builder/builder/src/main/java/io/helidon/pico/builder/BuilderTrigger.java [hunk 3]
```diff
@@ -25,7 +25,7 @@
 /**
  * the meta-annotation will trigger the enclosing service type marked with this annotation to drive the
```
[the remark this answers is not in the document]
Comment (trentjeff) [reply pico/builder/builder/src/main/java/io/helidon/pico/builder/BuilderTrigger.java:26]: done; will be in next push.

File: pico/builder/builder/src/test/java/io/helidon/pico/builder/spi/BeanUtilsTest.java [hunk 4]
```diff
@@ -53,96 +52,95 @@ void testIsValidMethodType() {
...
     void testValidateAndParseMethodName() {
-        AtomicReference<List<String>> attrName = new AtomicReference<>(Collections.emptyList());
+        AtomicReference<Optional<List<String>>> attrName = new AtomicReference<>();
 
         RuntimeException e = assertThrows(RuntimeException.class,
-                      () -> validateAndParseMethodName("x", "", true, attrName));
+                                          () -> validateAndParseMethodName("x", "", true, attrName));
         assertThat(e.getMessage(), is("invalid return type: x"));
-        assertThat(attrName.get(), nullValue());
+        assertThat(attrName.get(), is(Optional.empty()));
```
[the remark this answers is not in the document]
Comment (trentjeff) [reply pico/builder/builder/src/test/java/io/helidon/pico/builder/spi/BeanUtilsTest.java:67]: done; will be in next push.

File: pico/builder/tests/builder/src/test/java/io/helidon/pico/builder/api/test/UsageTest.java (hunk 2 above)
Comment (trentjeff) [reply pico/builder/tests/builder/src/test/java/io/helidon/pico/builder/api/test/UsageTest.java:40]: done; will be in next push.

File: pico/builder/tests/builder/src/test/java/io/helidon/pico/builder/api/test/MyConfigBeanTest.java (hunk 1 above)
Comment (trentjeff) [reply pico/builder/tests/builder/src/test/java/io/helidon/pico/builder/api/test/MyConfigBeanTest.java:42]: discussed this w/ Tomas. marking resolved for now - will create a task to do this later.
